go/private/go_repository.bzl

```diff
@@ -25,67 +25,84 @@ def _go_repository_impl(ctx):
+        type = ctx.attr.type,
+    )
+  else:
+    # Using fetch repo
+    if ctx.attr.vcs and not ctx.attr.remote:
+      fail("if vcs is specified, remote must also be")
+    # TODO(yugui): support submodule?
+    # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
+    result = ctx.execute([
+        ctx.path(ctx.attr._fetch_repo),
+        '--dest', ctx.path(''),
+        '--remote', ctx.attr.remote,
+        '--rev', rev,
+        '--vcs', ctx.attr.vcs,
+        '--importpath', ctx.attr.importpath,
+    ])
+    if result.return_code:
+      fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
 
+  if (ctx.attr.disable_build_file_generation or 
+      ctx.path('BUILD').exists or 
+      ctx.path(ctx.attr.build_file_name).exists):
+    # Do not generate build files for this repository
+    return
 
-def _new_go_repository_impl(ctx):
-  _go_repository_impl(ctx)
+  # Build file generation is needed
   gazelle = ctx.path(ctx.attr._gazelle)
-
   cmds = [gazelle, '--go_prefix', ctx.attr.importpath, '--mode', 'fix',
           '--repo_root', ctx.path(''),
           "--build_tags", ",".join(ctx.attr.build_tags)]
   if ctx.attr.build_file_name:
       cmds += ["--build_file_name", ctx.attr.build_file_name]
-
   cmds += [ctx.path('')]
-
   result = ctx.execute(cmds)
   if result.return_code:
-    fail("failed to generate BUILD files for %s: %s" % (
-        ctx.attr.importpath, result.stderr))
-
-
-_go_repository_attrs = {
-    "build_file_name": attr.string(default="BUILD.bazel"),
-    "importpath": attr.string(),
-    "remote": attr.string(),
-    "vcs": attr.string(default="", values=["", "git", "hg", "svn", "bzr"]),
-    "commit": attr.string(),
-    "tag": attr.string(),
-    "build_tags": attr.string_list(),
-    "_fetch_repo": attr.label(
-        default = Label("@io_bazel_rules_go_repository_tools//:bin/fetch_repo"),
-        allow_files = True,
-        single_file = True,
-        executable = True,
-        cfg = "host",
-    ),
-}
+      fail("failed to generate BUILD files for %s: %s" % (
+          ctx.attr.importpath, result.stderr))
 
 
 go_repository = repository_rule(
     implementation = _go_repository_impl,
-    attrs = _go_repository_attrs,
-)
+    attrs = {
+        # Fundamental attributes of a go repository
+        "importpath": attr.string(mandatory = True),
+        "commit": attr.string(),
+        "tag": attr.string(),
+        "build_tags": attr.string_list(),
 
+        # Attributes for a repository that cannot be inferred from the import path
+        "vcs": attr.string(default="", values=["", "git", "hg", "svn", "bzr"]),
+        "remote": attr.string(),
 
-new_go_repository = repository_rule(
-    implementation = _new_go_repository_impl,
-    attrs = _go_repository_attrs + {
+        # Attributes for a repository that comes from a source blob not a vcs
+        "url": attr.string(),
+        "strip_prefix": attr.string(),
```

Comment:
I'm really hesitant to encourage the use of `strip_prefix`, since renaming a GitHub repository can retroactively break everything that depends on it. I wonder if we should provide an alternative, maybe `strip_count` that removes the first n directory components? If we keep `strip_prefix`, we should definitely document the risk of using it.